ansible_rulebook/websocket.py

```diff
@@ -12,126 +12,208 @@
+from ansible_rulebook.auth import renew_token
 from ansible_rulebook.common import StartupArgs
+from ansible_rulebook.conf import settings
+from ansible_rulebook.exception import ShutdownException
 
 logger = logging.getLogger(__name__)
 
 
-async def request_workload(
-    activation_id: str, websocket_address: str, websocket_ssl_verify: str
-) -> StartupArgs:
-    logger.info("websocket %s connecting", websocket_address)
-    async with websockets.connect(
-        websocket_address,
-        ssl=_sslcontext(websocket_address, websocket_ssl_verify),
-    ) as websocket:
-        try:
-            logger.info("websocket %s connected", websocket_address)
-            await websocket.send(
-                json.dumps(dict(type="Worker", activation_id=activation_id))
-            )
+BACKOFF_MIN = 1.92
+BACKOFF_MAX = 60.0
+BACKOFF_FACTOR = 1.618
+BACKOFF_INITIAL = 5
+
 
-            project_data_fh = None
-            response = StartupArgs()
-            while True:
-                msg = await websocket.recv()
-                data = json.loads(msg)
-                if data.get("type") == "EndOfResponse":
+async def _connect_websocket(
+    handler: tp.Callable[[WebSocketClientProtocol], tp.Awaitable],
+    retry: bool,
+    **kwargs: list,
+) -> tp.Any:
+    logger.info("websocket %s connecting", settings.websocket_address)
+    if settings.websocket_access_token:
+        extra_headers = {
+            "Authorization": f"Bearer {settings.websocket_access_token}"
+        }
+    else:
+        extra_headers = {}
+
+    result = None
+
+    while True:
+        backoff_delay = BACKOFF_MIN
+        try:
+            async with websockets.connect(
+                settings.websocket_address,
+                ssl=_sslcontext(),
+                extra_headers=extra_headers,
+            ) as websocket:
+                result = await handler(websocket, **kwargs)
+                if not retry:
                     break
-                if data.get("type") == "ProjectData":
-                    if not project_data_fh:
-                        (
-                            project_data_fh,
-                            response.project_data_file,
-                        ) = tempfile.mkstemp()
-
-                    if data.get("data") and data.get("more"):
-                        os.write(
-                            project_data_fh, base64.b64decode(data.get("data"))
-                        )
-                    if not data.get("data") and not data.get("more"):
-                        os.close(project_data_fh)
-                        logger.debug("wrote %s", response.project_data_file)
-                if data.get("type") == "Rulebook":
-                    response.rulesets = rules_parser.parse_rule_sets(
-                        yaml.safe_load(base64.b64decode(data.get("data")))
+        except asyncio.CancelledError:  # pragma: no cover
+            raise
+        except ShutdownException:
+            break
+        except Exception as e:
```

Comment:
@bzwei can we catch websockets.exceptions.InvalidStatusCode here instead of catching Exception